postgres-testing/src/main/java/org/folio/postgres/testing/PostgresTesterContainer.java

```diff
@@ -34,44 +44,143 @@ public PostgresTesterContainer() {
    * Start the container.
    */
   @Override
-  public void start(String database, String username, String password) {
-    if (postgreSQLContainer != null) {
+  public void start(String database, String username, String password) throws PostgresTesterStartException {
+    if (primary != null) {
       throw new IllegalStateException("already started");
     }
-    postgreSQLContainer = new PostgreSQLContainer<>(dockerImageName)
-        .withDatabaseName(database)
-        .withUsername(username)
-        .withPassword(password)
-        .withStartupTimeout(Duration.ofSeconds(60));
-    postgreSQLContainer.start();
+
+    GenericContainer<?> tempStandby;
+    var replicationUser = "replicator";
+    var primaryHost = "primaryhost";
+    var dataDirectory = "/var/lib/postgresql/standby/";
+    var tempDirectory = "/tmp/standby/";
+    var hostVolume = "/tmp/rmb-standby-" + UUID.randomUUID();
+    var network = Network.newNetwork();
+
+    // In postgres replication the default is for commit to be async. So no configuration is needed, but our
+    // default for tests is to configure it for synchronous commit.
+    boolean configureSynchronousCommit = System.getProperty(POSTGRES_ASYNC_COMMIT) == null;
+
+    try {
+      primary = new PostgreSQLContainer<>(dockerImageName)
+          .withDatabaseName(database)
+          .withUsername(username)
+          .withPassword(password)
+          .withNetwork(network)
+          .withNetworkAliases(primaryHost)
+          .waitingFor(Wait.forLogMessage(".*database system is ready to accept connections.*\\n", READY_MESSAGE_TIMES));
+
+      if (configureSynchronousCommit) {
+        primary.withEnv("PGOPTIONS", "-c synchronous_commit=remote_apply");
+      }
+
+      primary.start();
+
+      // Modify_hba.conf to allow for replication.
+      var hbaConf = String.format("host replication %s 0.0.0.0/0 trust%n", replicationUser);
+      var echo = String.format("echo '%s' >> /var/lib/postgresql/data/pg_hba.conf", hbaConf);
+      primary.execInContainer("sh", "-c", echo);
+
+      // Reload primary configuration to allow for standby to connect to primary.
+      primary.execInContainer("psql", "-U", username, "-d", database, "-c", "SELECT pg_reload_conf();");
+      var waitForHbaRelooad = Wait.forLogMessage(".*database system is ready to accept connections.*", 1);
+      primary.waitingFor(waitForHbaRelooad);
+
+      // Create replication user.
+      var createReplicationUser = "CREATE USER " + replicationUser + " WITH REPLICATION PASSWORD 'abc123'";
+      primary.execInContainer("psql", "-U", username, "-d", database, "-c", createReplicationUser);
+
+      // Make a temporary container that only gets used to generate the data directory, which we bind to the host
+      // file system. Don't use PostgreSQLContainer for this because it will start a db and make it very hard to clear
+      // out the data dir which is required for pg_basebackup.
```

Comment:
The problems that the temp container solves don't have to do with running pg_basebackup per se, even though were running it there. The temp container solves two problems that are really related to using TestContainers for what we're doing. These "problems" are:
* We need run pg_basebackup in  `GenericContainer` rather than the `PostgresSQLContainer` because the latter starts pg, and effectively locks the data directory because it starts pg. We need a container that has PG but doesn't start it, hence the generic one.
* Because we're using a `GenericContainer` and because pg isn't started, test containers won't expose any ports on the container, and even if we did it wouldn't start, because TestContainers really wants a process running at the port, otherwise it again won't start.

There could be a way to do it with two containers, but I spent a lot of time trying to make it work and the only workaround I could find was what we have with the filesysystem mount. Let's keep it.

Reply:
> Can you explain in more detail why the straightforward way doesn't work?
> 
> [master...RMB-938-container-read-write-split-julian](https://github.com/folio-org/raml-module-builder/compare/master...RMB-938-container-read-write-split-julian)

What you have does appear to work but you already know that I was trying to do everything by `execInContainer` and ran into trouble with perms which you have solved with a `Transferable` and the `fileMode` param, which I did not know about. But what is the magic of the `docker-entrypoint-initdb.d` directory?

Reply:
I see that docker-entrypoint-initdb.d is the standard place for db containers to put sql and such for Postgres and mysql. New to me.